### PR TITLE
Build local web ui

### PR DESCRIPTION
**Pull request by lionel-nj**
The content of this PR was migrated to #421 

**Summary:**

This PR provides support to develop a local web ui based on the existing gtfs validator using spring boot, and react/

**Expected behavior:** 

1. Run the SpringBoot application:
`./gradlew bootRun` -> will start by default on server port 8080,

To run the web app :
1. go to reactclient directory
2. `yarn start`

image

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Format the title like "Fix #<issue_number> - <short description of fix and changes>" (for example - "Fix #1111 - Check for null value before using field")
- [ ] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Comment by fabrice-v** [comment]
Cool, looking promising. As discussed, next steps before possibly merging to master

- [ ] restructure readme and add documentation regarding the code structure (added modules) as well as how to/setup fro a developer
- [ ] research and implement minimal tests
- [ ] modify GitHub workflow
  - [ ] build web related modules and code
  - [ ] run web related tests
  - [ ] packaging as a .jar

File: reactclient/src/components/tests/NiceButton.test.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import NiceButton from "../NiceButton";
+import renderer from 'react-test-renderer';
+
+describe('NiceButton', () => {
+    it('should be defined', () => {
+        expect(NiceButton).toBeDefined();
+    });
+
+    it('should render correctly', () => {
+        let mockOnClick = jest.fn();
```
Comment (fabrice-v) [inline reactclient/src/components/tests/NiceButton.test.js:27]: For the exercize, use beforeAll as discussed to share a single mock for both tests

File: README.md [hunk 2]
```diff
@@ -131,15 +161,22 @@ Some important modules:
 * [Use cases](usecase) - Business logic 
 * [Adapters](adapter) - Convertors (e.g., parsers and exporters)
 * [application/cli-app](application/cli-app) - The main command-line application
+* [application/spring-boot-app](application/spring-boot-app) - The spring boot implementation of the application
+* [Client](client) - The web layer used to interact with the [spring-boot-app](application/spring-boot-app)
```
Comment (fabrice-v) [inline README.md:165]: Client --> reactclient

File: application/cli-app/src/main/java/org/mobilitydata/gtfsvalidator/Main.java [hunk 3]
```diff
@@ -247,26 +246,16 @@ public static void main(String[] args) {
...
-        } else if (!Strings.isNullOrEmpty(executionParametersAsString) || args.length == 0) {
-            // true when no arguments are provided or when json configuration is provided
-            logger.info("Retrieving execution parameters from execution-parameters.json file");
-            return new DefaultConfig(executionParametersAsString, logger);
-        } else {
-            // true when only arguments are provided
-            logger.info("Retrieving execution parameters from command-line");
-            return new DefaultConfig(args, logger);
-        }
+        final DefaultConfig.Builder configBuilder = new DefaultConfig.Builder();
```
Comment (fabrice-v) [inline application/cli-app/src/main/java/org/mobilitydata/gtfsvalidator/Main.java:255]: Nice improvement! 👍

File: application/spring-boot-app/build.gradle [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    implementation(
+            project(':config'),
+            project(':reactclient'),
```
Comment (fabrice-v) [inline application/spring-boot-app/build.gradle:42]: Is that a typo or why would the client be required to compile the server part? (Or am I getting something wrong here)

File: application/spring-boot-app/src/test/java/org/mobilitydata/gtfsvalidator/GtfsValidatorControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+@AutoConfigureMockMvc
+class GtfsValidatorControllerTest {
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private ServiceManager mockServiceManager;
+
+//    @Test
```
Comment (fabrice-v) [inline application/spring-boot-app/src/test/java/org/mobilitydata/gtfsvalidator/GtfsValidatorControllerTest.java:43]: Add a `//FIXME` explaining the intent or delete the commented section of the code

File: reactclient/jest.config.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+  // An array of regexp pattern strings used to skip coverage collection
+  // coveragePathIgnorePatterns: [
+  //   "/node_modules/"
+  // ],
+
+  // Indicates which provider should be used to instrument code for coverage
+  // coverageProvider: "v8",
+
+  // A list of reporter names that Jest uses when writing coverage reports
```
Comment (fabrice-v) [inline reactclient/jest.config.js:47]: We will have to study if that'd be relevant to integrate coverage testing in our test suite

**Review by fabrice-v — CHANGES_REQUESTED** [review]
We're almost there, very neat!
A few changes inline

File: application/spring-boot-app/build.gradle (hunk 4 above)
Comment (lionel-nj) [reply application/spring-boot-app/build.gradle:42]: We must have this dependency here to allow build of react layer

File: reactclient/src/components/tests/JsonRenderer.test.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import renderer from 'react-test-renderer';
+
+describe('JsonRenderer', () => {
+    it('should be defined without data', () => {
+        expect(JsonRenderer).toBeDefined();
+    });
+
+    it('should render correctly without data', () => {
+        const tree = renderer.create(
+            <JsonRenderer data={{}}/>
```
[the remark this answers is not in the document]
Comment (lionel-nj) [reply reactclient/src/components/tests/JsonRenderer.test.js:28]: Indeed, passing `null` data generates an error and it should be reflected in the tests

File: README.md [hunk 8]
```diff
@@ -131,15 +161,22 @@ Some important modules:
 * [Use cases](usecase) - Business logic 
 * [Adapters](adapter) - Convertors (e.g., parsers and exporters)
 * [application/cli-app](application/cli-app) - The main command-line application
+* [application/spring-boot-app](application/spring-boot-app) - The spring boot implementation of the application
+* [reactclient](reactclient) - The web layer used to interact with the [spring-boot-app](application/spring-boot-app)
```
Comment (fabrice-v) [inline README.md:165]: reactclient -> react-client

File: README.md [hunk 9]
```diff
@@ -131,15 +161,22 @@ Some important modules:
...
 * [Adapters](adapter) - Convertors (e.g., parsers and exporters)
 * [application/cli-app](application/cli-app) - The main command-line application
+* [application/spring-boot-app](application/spring-boot-app) - The spring boot implementation of the application
+* [reactclient](reactclient) - The web layer used to interact with the [spring-boot-app](application/spring-boot-app) 
 
 # Tests
 
 To run tests: 
-
+1. CLI and server tests
```
Comment (fabrice-v) [inline README.md:170]: It should have the same name as the workflow step

File: adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java [hunk 10]
```diff
@@ -16,7 +16,7 @@
 
 package org.mobilitydata.gtfsvalidator.exporter;
 
-import org.mobilitydata.gtfsvalidator.adapter.protos.GtfsValidationOutputProto;
+import org.mobilitydata.gtfsvalidator.protos.GtfsValidationOutputProto;
```
Comment (fabrice-v) [inline adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java:19]: As discussed, thos files should be moved back to their original location

File: adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java [hunk 11]
```diff
@@ -30,7 +30,7 @@
 import java.util.ArrayList;
 import java.util.List;
 
-import static org.mobilitydata.gtfsvalidator.adapter.protos.GtfsValidationOutputProto.GtfsProblem.Type.*;
+import static org.mobilitydata.gtfsvalidator.protos.GtfsValidationOutputProto.GtfsProblem.Type.*;
```
Comment (fabrice-v) [inline adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java:33]: Move back to original location

File: adapter/exporter/src/test/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporterTest.java [hunk 12]
```diff
@@ -17,7 +17,7 @@
 package org.mobilitydata.gtfsvalidator.exporter;
 
 import org.junit.jupiter.api.Test;
-import org.mobilitydata.gtfsvalidator.adapter.protos.GtfsValidationOutputProto;
+import org.mobilitydata.gtfsvalidator.protos.GtfsValidationOutputProto;
```
Comment (fabrice-v) [inline adapter/exporter/src/test/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporterTest.java:20]: Move back to original location

File: adapter/exporter/src/test/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporterTest.java [hunk 13]
```diff
@@ -29,7 +29,7 @@
 import java.time.LocalDate;
 import java.util.List;
 
-import static org.mobilitydata.gtfsvalidator.adapter.protos.GtfsValidationOutputProto.GtfsProblem.Type.*;
+import static org.mobilitydata.gtfsvalidator.protos.GtfsValidationOutputProto.GtfsProblem.Type.*;
```
Comment (fabrice-v) [inline adapter/exporter/src/test/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporterTest.java:32]: Move back to original location

File: adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryValidationResultRepository.java [hunk 14]
```diff
@@ -17,7 +17,7 @@
 package org.mobilitydata.gtfsvalidator.db;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
-import org.mobilitydata.gtfsvalidator.adapter.protos.GtfsValidationOutputProto;
+import org.mobilitydata.gtfsvalidator.protos.GtfsValidationOutputProto;
```
Comment (fabrice-v) [inline adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryValidationResultRepository.java:20]: Move back to original location

File: reactclient/src/components/tests/JsonRenderer.test.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (lionel-nj) [reply reactclient/src/components/tests/JsonRenderer.test.js:28]: Test to be implemented, I added a FIXME tag to this intent

**Review by fabrice-v — CHANGES_REQUESTED** [review]
A few changes still needed. See inline

**Comment by lionel-nj** [comment]
This PR will be closed without merger. To follow-up the work done in, please refer to https://github.com/MobilityData/gtfs-validator/pull/421.

**Comment by fabrice-v** [comment]
Content was migrated and merged through #421

**Pull request closed without merging**